### PR TITLE
Geo Indexing (WIP)

### PR DESCRIPTION
**Pull request by kishoreg**
## Description

Adding support for indexing (lat,lon) using Uber's H3 indexing.

```
select count(*) from starbucksStores  where ST_DISTANCE(location_st_point , ST_POINT(-122, 37)) < 10
```

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  private final Set<ColumnMetadata> _h3IndexColumns = new HashSet<>();
+
+  public H3IndexHandler(File indexDir, SegmentMetadataImpl segmentMetadata, IndexLoadingConfig indexLoadingConfig,
+      SegmentDirectory.Writer segmentWriter) {
+    _indexDir = indexDir;
+    _segmentWriter = segmentWriter;
+    _segmentName = segmentMetadata.getName();
+    _segmentVersion = SegmentVersion.valueOf(segmentMetadata.getVersion());
+
+    // Only create H3 index on non-dictionary-encoded columns
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java:68]: what does this mean?

Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java:68]: This actually brings up an interesting question. If user wants to store the point with column lat, lng as double, is it possible to create an index for them?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/geospatial/H3IndexReader.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      return new MutableRoaringBitmap();
+    }
+    // Return the bitmap if it's still on heap
+    if (_bitmaps != null) {
+      bitmapArrayReference = _bitmaps.get();
+      if (bitmapArrayReference != null) {
+        SoftReference<ImmutableRoaringBitmap> bitmapReference = bitmapArrayReference[dictId];
+        if (bitmapReference != null) {
+          ImmutableRoaringBitmap value = bitmapReference.get();
+          if (value != null) {
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/geospatial/H3IndexReader.java:80]: can the value be null? what does that mean

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java [hunk 3]
```diff
@@ -72,4 +73,17 @@ public static String stAsText(byte[] bytes) {
...
   }
+
+  /**
+   * Saves the geometry object as WKT format.
+   *
+   * @param bytes the serialized geometry object
+   * @return the geometry in WKT
+   */
+  @ScalarFunction
+  public static byte[] toSphericalGeography(byte[] bytes) {
```
[the remark this answers is not in the document]
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:84]: this is what Presto had

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java [hunk 4]
```diff
@@ -325,10 +326,10 @@ public long getLatestIngestionTimestamp() {
       // Null value vector
       MutableNullValueVector nullValueVector = _nullHandlingEnabled ? new MutableNullValueVector() : null;
 
-      // TODO: Support range index and bloom filter for mutable segment
+      // TODO: Support range index and bloom filter and h3 index for mutable segment
       _indexContainerMap.put(column,
           new IndexContainer(fieldSpec, partitionFunction, partitions, new NumValuesInfo(), forwardIndex, dictionary,
-              invertedIndexReader, null, textIndex, null, nullValueVector));
+              invertedIndexReader, null, null, textIndex, null, nullValueVector));
```
[the remark this answers is not in the document]
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java:332]: yes, but should be part of another PR

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  private final H3Core _h3Core;
+  private final H3IndexReader _h3IndexReader;
+  private Geometry _geometry;
+  private double _distance;
+
+  public H3IndexFilterOperator(Predicate predicate, IndexSegment indexSegment, int numDocs) {
+    FunctionContext function = predicate.getLhs().getFunction();
+    String columnName;
+
+    if (function.getArguments().get(0).getType() == ExpressionContext.Type.IDENTIFIER) {
```
[the remark this answers is not in the document]
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java:58]: Not yet but its possible to handle them

File: pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java [hunk 6]
```diff
@@ -120,6 +129,20 @@ private BaseFilterOperator constructPhysicalOperator(FilterContext filter,
         Predicate predicate = filter.getPredicate();
         ExpressionContext lhs = predicate.getLhs();
         if (lhs.getType() == ExpressionContext.Type.FUNCTION) {
+          FunctionContext function = lhs.getFunction();
+
+          boolean canApplyH3Index = false;
+          if (function.getFunctionName().equalsIgnoreCase(StDistanceFunction.FUNCTION_NAME)) {
+            String columnName = function.getArguments().get(0).getIdentifier();
```
[the remark this answers is not in the document]
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:136]: no, it means column name is the first part of the argument to st_distance function. I thought I handled both cases, let me check

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/geospatial/H3IndexCreator.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+  int numChunks = 0;
+  List<Integer> chunkLengths = new ArrayList<>();
+
+  public H3IndexCreator(File indexDir, FieldSpec fieldSpec, int resolution)
+      throws IOException {
+
+    _indexDir = indexDir;
+    _fieldSpec = fieldSpec;
+    _resolution = resolution;
```
[the remark this answers is not in the document]
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/geospatial/H3IndexCreator.java:74]: we should have that as part of the indexconfig. May be support multiple resolutions. No need to extend this class.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java (hunk 1 above)
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/H3IndexHandler.java:68]: this is for POC, dictionary encoded should also be possible. If user has lat, lon - then one should use derived virtual column (TBD) to create a geom column and generate h3 index on that.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/geospatial/H3IndexReader.java (hunk 2 above)
Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/geospatial/H3IndexReader.java:80]: it means the reference was garbage collected

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java [hunk 8]
```diff
@@ -157,11 +157,11 @@ static Object standardize(String column, @Nullable Object value, boolean isSingl
       if (numStandardizedValues == 0) {
         return null;
       }
-      if (numStandardizedValues == 1) {
+      if (numStandardizedValues == 1 || isSingleValue) {
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java:160]: Is this needed in this PR?

**Comment by Jackie-Jiang** [comment]
Switched to #6409

**Pull request closed without merging**